src/compiler/scala/tools/nsc/Global.scala

```diff
@@ -75,16 +78,16 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
 
   override def settings = currentSettings
 
-  private[this] var currentReporter: Reporter = { reporter = reporter0 ; currentReporter }
-
-  def reporter: Reporter = currentReporter
+  private[this] val currentReporter: WorkerOrMainThreadLocal[Reporter] =
+    WorkerThreadLocal(new BufferedReporter, reporter0)
+  def reporter: Reporter =  currentReporter.get
   def reporter_=(newReporter: Reporter): Unit =
-    currentReporter = newReporter match {
+    currentReporter.set(newReporter match {
```

Comment:
assert Parallel.onMainThread?

Reply:
Wouldn't be true.
We are setting new one for every unit now.
It's required to ensure consistent ordering of the logs.